californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -396,11 +398,12 @@ public void completed(final Exchange exchange) {
 
 		@Override
 		public void contextEstablished(final Exchange exchange) {
-
-			if (exchange.getRequest() != null) {
-				observationStore.setContext(exchange.getRequest().getToken(), exchange.getCorrelationContext());
+			Request currentRequest = exchange.getCurrentRequest(); 
+			Request request = exchange.getRequest(); 
+			if (request == currentRequest && request != null && request.isObserve()) {
+				observationStore.setContext(request.getToken(), exchange.getCorrelationContext());
```

Comment:
> For me it's more important to know, which cases would require a setContext XyzStorage call. For now, it seems to be observe. Are there other use cases?

What do you call `XyzStorage` ? we need context for all Request/Response correlation (observe or not), so for all exchanges.

> Why do we need setContext in both storages? 

Both storages (ObservationStore and MessageExchangeStore) store exchanges. When a correlation context is established we must associate it to an exchange. So it should be stored (persisted for persistent store) that's why there is this `setContext` method. For now, I think there is only persistent implementation of `ObservationStore`, as far as I know `MessageExchangeStore` exist only in memory.  

> Why is one called for the request, the other one for currentRequest? 

About `ObservationStore`, we only store observe exchange (we don't store magical blockwise exchange) so we are only interested by the `request` (as `currentrequest` is about blockwise).

About `MessageExchangeStore`, we store every thing except "long term exchange relation" so in that case we care about blockwise transfert so I suppose this is the idea about why we use `currentRequest` instead of `request`.

FMPOV, setting correlationContext on observation store only for observe request make sense to me. (good optimization)
Without thinking too much , raising the event `onCorrelationContext` only when it really change can not be a bad modification. (using equals not != of course :p)

But if we think more about that, Exchange is used for :
 - Request/Response.
 - ObserverRequest/Response/notification
 - Resquest/Response + magical blockwise
For me in all case, the correlation context should not changed.
With our current implementation, the only case where context is changed (I mean not initially set) is when we do blockwise transfer. The question is "Should we allow to change context for a whole blockwise transfer?" 
Personally, I'll be tented to say "no".
It's not clear to me but the [RFC 7959](https://tools.ietf.org/html/rfc7959#section-7) seems to say "no" too :
_" Where access to a resource is only granted to clients making use of
   specific security associations, all blocks of that resource MUST be
   subject to the same security checks; "_